### ISSUE TITLE
Version 4 - Looking for Testers

### ISSUE THREAD

**Issue by SamKirkland** [SamKirkland/FTP-Deploy-Action#102]
# Version 4 🚀
Over the past few months I've been thinking about this problems and how to create a new architecture that can resolve them. The new version of this library is heavily inspired by `rsync`. `rsync` is a popular application that must be run over a `ssh` connection, it requires the `rsync` binary be install on both the client and the sever. `rsync` provides dozens of advanced arguments that can't be supported due to protocol limitations with `ftp`. Version 4 attempts to emulate at a high level how `rsync` syncs multiple folders, however operates over the constrained `ftp` protocol.

The goal of version 4 is to build a slim version of `rsync`. The complicated bit comes from working around the limitations of `ftp`, `v1` and `v2` taught me valuable lessons around `ftp`. The largest problem comes from diff based deployments, without them publishes must delete all server files, then re-deploy from a blank slate. This is trivial to program but causes deployments that can last several hours...

My solution with v4 is to build a `.ftp-deploy-sync-state.json` file. This file is a simple nested dir listing of every file/folder in the local branch. Files are hashed and then a diffing algorithm detects the necessary file changes needed on the server to make it match the client (`upload`, `replace`, `delete`, `folderAdd`, `folderRemove`) - note: renames are not currently supported for simplicity but could be added at a later date.

These diffs are then converted to `ftp` commands and sent to the server. A large amount of code exists around error handling to make sure the script is as forgiving as possible. Once the sync is complete the `.ftp-deploy-sync-state.json` file is uploaded to the server so we can calculate diffs on the next deployment

### Decouple from git
Version 3 is tightly coupled with git source control. This tight coupling allowed for extremely fast publishes because only differences are synced to the server. Git history is used to track which version is currently deployed. In hindsight this was a mistake, tightly coupling to git resulted in [complex configurations](https://github.com/SamKirkland/FTP-Deploy-Action/tree/3.1.1#force-upload-specific-files) for basic use cases. Untracked files were deployed every time, which negated many of the performance benefits git provided. Using git also caused hard-to-debug issues like the `fetch-depth` problem that still comes up in non-standard use cases. Publishing a folder dynamically generated from build scripts becomes difficult and cumbersome.

Version 4 will no longer use git in any way. It will still run via github actions and be able to publish git tracked files. Because git is no longer used as the diffing solution `.git-ftp-include` and `.git-ftp-ignore` are no longer needed, the script will publish all files/folders (I will provide glob filter options).

### Improved Defaults
`.git` and `node_modules` files should be ignored by default, allowing users to opt-in to their deployment instead.

### Improved Error Handling
Errors were cryptic. Because this action is based around third parties adding error messages is difficult and requires third party PRs and a long delay in order to see those changes in this down-stream library. Error messages should tell users exactly what went wrong and how to fix it.

### Local Testing
Testing actions today is difficult, often requiring devs to change a value and commit it without testing the change locally. Using docker makes testing actions a large amount of work. Allowing users to run this command from powershell/terminal with shorten the feedback loop

### Minimize downtime
Maintaining this code myself gives me the opportunity to sync changes in a way that minimizes downtime. For example, new files can be uploaded, then replace updated files, and finally delete old files.

## Removed features 🗑️
`sftp` support has been removed. In hindsight this never should have been added. Users with `sftp` access should be using `rsync` over `ssh`. I will create a separate github action for rsync deployments.

The important arguments have been migrated (`--dry-run`) but most `git-ftp` commands have been removed (`--silent`, `--all`, `--lock`, `--key`, `--branch`, `--pubkey`, `--insecure`, `--cacert`)

-----

### Open Issues Resolved in Version 4 ✅
- [x] Removing git-based file tracking #101, #99, #88, #82, #78, #76, #68, #56, #52, #49, #40
- [x] Local testing - #71
- [x] Better defaults - #66, #57
- [x] Better error handling - #48, #19
- [x] Detect when hosts require ftp/ftps/ftpes and output helper text - related issues #87
- [x] Other issues - #83
- [x] better documentation/warning on escaping spaces/quotes - #80

# Testers 🐞
I've been working on version 4 over the past several weeks. I feel comfortable enough at this point to bring on anyone in the community that would be willing to help test these changes. If you would like to help out please let me know (with a comment or direct message).

I'm also looking for someone with a host that requires secure ftp (ftps or ftpes) as my current hosting company does not require secure connections. If someone can provide me with a limited access ftp account to play around with, or knows a host that has this restriction so I can buy a account it would help.

**Comment by jayhybrid** [comment]
My hoster requires `ftps://` and I am willing to test version 4

**Comment by SamKirkland** [comment]
> My hoster requires `ftps://` and I am willing to test version 4

Awesome! Here is the beta branch --> https://github.com/SamKirkland/FTP-Deploy-Action/blob/beta-v4/README.md
I have a known bug with handling spaces or quotes in the password.

Let me know if you run into any problems, thanks for helping!

**Comment by devmount** [comment]
My hoster requires `ftpes://`, I'll let you know when I had some time to test v4 👍🏻

**Comment by mikedelgaudio** [comment]
Just tried it out with my provider Bluehost (cPanel) using FTPS but I don't believe they enforce FTPS. It did successfully upload but I discovered two issues trying to upload a compiled Angular project however, do you have any specific contribution guidelines?

**Comment by SamKirkland** [comment]
@mikedelgaudio you can make a github issue and I'll look into a fix

**Comment by ivandjuricic** [comment]
On Sep 14 (haven't tried since) 
I got error: `Mon Sep 14 21:26:02 UTC 2020: fatal: Unrecognised option: ***`
There is this issue already like this: https://github.com/SamKirkland/FTP-Deploy-Action/issues/83

My yml was:
```  
- name: FTP-Deploy-Action
        uses: SamKirkland/FTP-Deploy-Action@beta-v4
        with:
          ftp-server: ${{ secrets.FTP_HOST }}
          ftp-username: ${{ secrets.FTP_USER }}
          ftp-password: ${{ secrets.FTP_PASSWORD }}
          local-dir: "public"
          git-ftp-args: --all
```

**Comment by AlexMayol** [comment]
This version 4 is good news! I'm glad you're dropping git-based file tracking, it was too troublesome. I'll adopt v4 as soon as you release it. 
Thanks for your hard work <3

**Comment by Mdelaf** [comment]
This seems a great improvement. I really appreciate your effort. Do you have any release date in mind? 

By the way, do you think this new version would allow some way to tackle the problem I mentioned to you some time ago?

> Every time I build my Angular app, the generated files have different names because their names contain a hash associated to their content.
> 
> The problem with this is that on every deploy, more and more files accumulate on the server and they never get deleted nor replaced.

**Comment by SamKirkland** [comment]
@Mdelaf V4 should resolve that issue. I've tested it with my react build which also uses hash based file cache-busting. Please try it out if you get a chance!!!

Release date TBD, I don't want to commit to a date but I think we are getting close. Would love more testers but I might have to just go for it at this point.

**Comment by Mdelaf** [comment]
Great! I will give it a try in the upcoming days and let you know :+1:
